sympy/matrices/eigen.py

```diff
@@ -1357,6 +1362,142 @@ def pick_vec(small_basis, big_basis):
 
     return restore_floats2(basis_mat, jordan_mat)
 
+def jordan_form_rational_matrix(M, calc_transform):
+    """
+    Return (P, J), where J is the Jordan form and P satisfies M = P J P^{-1},
+    if the matrix is rational. If calc_transform is False, return only J.
+    """
+
+    # cache calculations for some speedup
+    char_mat_cache = {}
+
+    def char_mat(val, pow):
+        """Cache computations of ``(M - val*I)**pow`` for quick
+        retrieval"""
+
+        if (val, pow) in char_mat_cache:
+            return char_mat_cache[(val, pow)]
+
+        else:
+            eigen_scalar = a if a is not None else val
+            char_mat_cache[(val, pow)] = (M - eigen_scalar*M.eye(M.rows)).to_DM(extension=True).pow(pow)
+        return char_mat_cache[(val, pow)]
```

Comment:
I don't understand why this needs to be cached. Why would the same thing be computed twice?

Also if we want many powers like `A**1, A**2, A**3, ...` then it is better to compute them using a single loop with multiplication.